File: curator/github_publish.py
from __future__ import annotations

import base64
import os
from dataclasses import dataclass
from pathlib import Path
from typing import Any

import requests


REPO_ROOT = Path(__file__).resolve().parents[1]
GITHUB_API_BASE_URL = "https://api.github.com"


@dataclass(frozen=True)
class GitHubPublishConfig:
    token: str
    owner: str
    repo: str
    branch: str


def _get_streamlit_secret(name: str) -> str:
    try:
        import streamlit as st

        value = st.secrets.get(name, "")
        return str(value).strip()
    except Exception:
        return ""


def _get_secret(name: str, default: str = "") -> str:
    return (
        os.environ.get(name, "").strip()
        or _get_streamlit_secret(name)
        or default
    ).strip()


def get_github_publish_config() -> GitHubPublishConfig:
    token = _get_secret("GITHUB_TOKEN")
    owner = _get_secret("GITHUB_REPO_OWNER")
    repo = _get_secret("GITHUB_REPO_NAME")
    branch = _get_secret("GITHUB_BRANCH", "main")

    missing = []

    if not token:
        missing.append("GITHUB_TOKEN")

    if not owner:
        missing.append("GITHUB_REPO_OWNER")

    if not repo:
        missing.append("GITHUB_REPO_NAME")

    if missing:
        raise RuntimeError(
            "Missing GitHub publish configuration: "
            + ", ".join(missing)
            + ". Set them as environment variables or Streamlit secrets."
        )

    return GitHubPublishConfig(
        token=token,
        owner=owner,
        repo=repo,
        branch=branch or "main",
    )


def get_github_config_summary() -> str:
    token_present = bool(_get_secret("GITHUB_TOKEN"))
    owner = _get_secret("GITHUB_REPO_OWNER")
    repo = _get_secret("GITHUB_REPO_NAME")
    branch = _get_secret("GITHUB_BRANCH", "main")

    return "\n".join(
        [
            f"GITHUB_TOKEN present: {token_present}",
            f"GITHUB_REPO_OWNER: {owner or '[missing]'}",
            f"GITHUB_REPO_NAME: {repo or '[missing]'}",
            f"GITHUB_BRANCH: {branch or '[missing]'}",
        ]
    )


class GitHubContentsClient:
    def __init__(self, config: GitHubPublishConfig):
        self.config = config

    @property
    def headers(self) -> dict[str, str]:
        return {
            "Accept": "application/vnd.github+json",
            "Authorization": f"Bearer {self.config.token}",
            "X-GitHub-Api-Version": "2022-11-28",
        }

    def _contents_url(self, repo_path: str) -> str:
        clean_path = repo_path.strip().lstrip("/")
        return (
            f"{GITHUB_API_BASE_URL}/repos/"
            f"{self.config.owner}/{self.config.repo}/contents/{clean_path}"
        )

    def get_file_info(self, repo_path: str) -> dict[str, Any] | None:
        response = requests.get(
            self._contents_url(repo_path),
            headers=self.headers,
            params={"ref": self.config.branch},
            timeout=30,
        )

        if response.status_code == 404:
            return None

        if not response.ok:
            raise RuntimeError(
                f"GitHub GET failed for {repo_path}: "
                f"{response.status_code} {response.text[:500]}"
            )

        payload = response.json()

        if payload.get("type") != "file":
            raise RuntimeError(f"GitHub path exists but is not a file: {repo_path}")

        return payload

    def get_existing_file_bytes(self, file_info: dict[str, Any]) -> bytes | None:
        content = str(file_info.get("content", "") or "")
        encoding = str(file_info.get("encoding", "") or "")

        if not content or encoding != "base64":
            return None

        compact_content = content.replace("\n", "")
        return base64.b64decode(compact_content.encode("utf-8"))

    def put_file(
        self,
        local_path: Path,
        repo_path: str,
        commit_message: str,
    ) -> dict[str, Any]:
        local_path = Path(local_path)

        if not local_path.exists():
            raise FileNotFoundError(f"Local file does not exist: {local_path}")

        file_bytes = local_path.read_bytes()
        file_info = self.get_file_info(repo_path)

        sha = None
        action = "created"

        if file_info is not None:
            sha = str(file_info.get("sha", "") or "")
            existing_bytes = self.get_existing_file_bytes(file_info)

            if existing_bytes == file_bytes:
                return {
                    "ok": True,
                    "path": repo_path,
                    "action": "skipped",
                    "message": f"No GitHub update needed for {repo_path}; content is unchanged.",
                    "html_url": str(file_info.get("html_url", "") or ""),
                }

            action = "updated"

        encoded_content = base64.b64encode(file_bytes).decode("utf-8")

        payload: dict[str, Any] = {
            "message": commit_message,
            "content": encoded_content,
            "branch": self.config.branch,
        }

        if sha:
            payload["sha"] = sha

        response = requests.put(
            self._contents_url(repo_path),
            headers=self.headers,
            json=payload,
            timeout=30,
        )

        if not response.ok:
            raise RuntimeError(
                f"GitHub PUT failed for {repo_path}: "
                f"{response.status_code} {response.text[:1000]}"
            )

        result = response.json()

        return {
            "ok": True,
            "path": repo_path,
            "action": action,
            "message": f"{repo_path} {action} on GitHub.",
            "html_url": str(result.get("content", {}).get("html_url", "") or ""),
            "commit_sha": str(result.get("commit", {}).get("sha", "") or ""),
        }


def repo_path_from_local_path(local_path: Path) -> str:
    local_path = Path(local_path).resolve()

    try:
        return local_path.relative_to(REPO_ROOT).as_posix()
    except ValueError as exc:
        raise ValueError(
            f"File must be inside repository root. File: {local_path}; repo root: {REPO_ROOT}"
        ) from exc


def publish_files_to_github(
    live_path: str | Path,
    batch_path: str | Path,
    commit_message: str,
    extra_paths: list[str | Path] | None = None,
) -> dict[str, Any]:
    config = get_github_publish_config()
    client = GitHubContentsClient(config)

    publish_paths = [
        Path(live_path),
        Path(batch_path),
    ]

    if extra_paths:
        for extra_path in extra_paths:
            publish_paths.append(Path(extra_path))

    # Avoid duplicate uploads while preserving order.
    unique_publish_paths: list[Path] = []
    seen_paths: set[str] = set()

    for path in publish_paths:
        resolved_key = str(path.resolve()).lower()

        if resolved_key in seen_paths:
            continue

        seen_paths.add(resolved_key)
        unique_publish_paths.append(path)

    uploads = []

    for local_path in unique_publish_paths:
        repo_path = repo_path_from_local_path(local_path)

        upload_result = client.put_file(
            local_path=local_path,
            repo_path=repo_path,
            commit_message=commit_message,
        )

        uploads.append(upload_result)

    output_lines = [
        f"Repository: {config.owner}/{config.repo}",
        f"Branch: {config.branch}",
        "",
    ]

    for upload in uploads:
        output_lines.append(
            f"- {upload['path']}: {upload['action']}"
        )

        if upload.get("html_url"):
            output_lines.append(f"  {upload['html_url']}")

        if upload.get("commit_sha"):
            output_lines.append(f"  commit: {upload['commit_sha']}")

    changed_count = sum(
        1 for upload in uploads
        if upload.get("action") in {"created", "updated"}
    )

    skipped_count = sum(
        1 for upload in uploads
        if upload.get("action") == "skipped"
    )

    return {
        "ok": True,
        "changed_count": changed_count,
        "skipped_count": skipped_count,
        "uploads": uploads,
        "message": (
            f"GitHub API publish completed. "
            f"Changed files: {changed_count}; unchanged files skipped: {skipped_count}."
        ),
        "output": "\n".join(output_lines),
    }

def publish_file_to_github(
    local_path: str | Path,
    commit_message: str | None = None,
    repo_path: str | None = None,
) -> dict[str, Any]:
    config = get_github_publish_config()
    client = GitHubContentsClient(config)

    local_path = Path(local_path)

    if repo_path is None:
        repo_path = repo_path_from_local_path(local_path)

    if commit_message is None:
        commit_message = f"Upload {repo_path}"

    upload_result = client.put_file(
        local_path=local_path,
        repo_path=repo_path,
        commit_message=commit_message,
    )

    output_lines = [
        f"Repository: {config.owner}/{config.repo}",
        f"Branch: {config.branch}",
        "",
        f"- {upload_result['path']}: {upload_result['action']}",
    ]

    if upload_result.get("html_url"):
        output_lines.append(f"  {upload_result['html_url']}")

    if upload_result.get("commit_sha"):
        output_lines.append(f"  commit: {upload_result['commit_sha']}")

    return {
        "ok": True,
        "changed_count": 0 if upload_result.get("action") == "skipped" else 1,
        "upload": upload_result,
        "message": f"GitHub PDF upload completed: {repo_path} {upload_result['action']}.",
        "output": "\n".join(output_lines),
    }

File: src/github-publish.js
function cleanText(
  value
) {
  return String(
    value ?? ""
  ).trim();
}


function getGitHubConfig(
  env
) {
  return {
    token:
      cleanText(
        env.GITHUB_TOKEN
      ),

    owner:
      cleanText(
        env.GITHUB_REPO_OWNER ||
        "cbixx"
      ),

    repo:
      cleanText(
        env.GITHUB_REPO_NAME ||
        "global-corrosion-site-maps"
      ),

    branch:
      cleanText(
        env.GITHUB_BRANCH ||
        "main"
      ),
  };
}


function requireGitHubConfig(
  env
) {
  const config =
    getGitHubConfig(
      env
    );

  const missing = [];

  if (!config.token) {
    missing.push(
      "GITHUB_TOKEN"
    );
  }

  if (!config.owner) {
    missing.push(
      "GITHUB_REPO_OWNER"
    );
  }

  if (!config.repo) {
    missing.push(
      "GITHUB_REPO_NAME"
    );
  }

  if (
    missing.length
  ) {
    throw new Error(
      "Missing GitHub publish configuration: " +
      missing.join(", ")
    );
  }

  return config;
}


function githubHeaders(
  config
) {
  return {
    accept:
      "application/vnd.github+json",

    authorization:
      `Bearer ${config.token}`,

    "x-github-api-version":
      "2022-11-28",
  };
}


function encodeRepoPath(
  repoPath
) {
  return String(
    repoPath
  )
    .replace(
      /^\/+/,
      ""
    )
    .split("/")
    .map(
      (part) =>
        encodeURIComponent(
          part
        )
    )
    .join("/");
}


function contentsUrl(
  config,
  repoPath
) {
  return (
    `https://api.github.com/repos/` +
    `${encodeURIComponent(config.owner)}/` +
    `${encodeURIComponent(config.repo)}/` +
    `contents/${encodeRepoPath(repoPath)}`
  );
}


function encodeUtf8Base64(
  text
) {
  const bytes =
    new TextEncoder()
      .encode(
        text
      );

  let binary = "";

  const chunkSize =
    0x8000;

  for (
    let offset = 0;
    offset < bytes.length;
    offset += chunkSize
  ) {
    const chunk =
      bytes.subarray(
        offset,
        Math.min(
          offset +
            chunkSize,
          bytes.length
        )
      );

    binary +=
      String.fromCharCode(
        ...chunk
      );
  }

  return btoa(
    binary
  );
}


function decodeBase64Bytes(
  value
) {
  const compact =
    String(
      value ||
      ""
    )
      .replace(
        /\s+/g,
        ""
      );

  if (!compact) {
    return null;
  }

  const binary =
    atob(
      compact
    );

  const bytes =
    new Uint8Array(
      binary.length
    );

  for (
    let index = 0;
    index < binary.length;
    index += 1
  ) {
    bytes[index] =
      binary.charCodeAt(
        index
      );
  }

  return bytes;
}


function bytesEqual(
  first,
  second
) {
  if (
    first.length !==
    second.length
  ) {
    return false;
  }

  for (
    let index = 0;
    index < first.length;
    index += 1
  ) {
    if (
      first[index] !==
      second[index]
    ) {
      return false;
    }
  }

  return true;
}


async function getGitHubFileInfo(
  config,
  repoPath
) {
  const url =
    new URL(
      contentsUrl(
        config,
        repoPath
      )
    );

  url.searchParams.set(
    "ref",
    config.branch
  );

  const response =
    await fetch(
      url,
      {
        headers:
          githubHeaders(
            config
          ),
      }
    );

  if (
    response.status ===
    404
  ) {
    return null;
  }

  if (!response.ok) {
    const detail =
      await response.text();

    throw new Error(
      `GitHub GET failed for ${repoPath}: ` +
      `${response.status} ${detail.slice(0, 400)}`
    );
  }

  const payload =
    await response.json();

  if (
    payload.type !==
    "file"
  ) {
    throw new Error(
      `GitHub path exists but is not a file: ${repoPath}`
    );
  }

  return payload;
}


async function getExistingGitHubBytes(
  config,
  fileInfo
) {
  if (
    fileInfo.encoding ===
      "base64" &&
    fileInfo.content
  ) {
    return decodeBase64Bytes(
      fileInfo.content
    );
  }

  const downloadUrl =
    cleanText(
      fileInfo.download_url
    );

  if (!downloadUrl) {
    return null;
  }

  const response =
    await fetch(
      downloadUrl,
      {
        headers: {
          accept:
            "application/octet-stream",
        },
      }
    );

  if (!response.ok) {
    return null;
  }

  return new Uint8Array(
    await response.arrayBuffer()
  );
}


async function putGitHubFile(
  config,
  {
    repoPath,
    content,
    commitMessage,
  }
) {
  const fileInfo =
    await getGitHubFileInfo(
      config,
      repoPath
    );

  const newBytes =
    new TextEncoder()
      .encode(
        content
      );

  let action =
    "created";

  let sha =
    "";

  if (fileInfo) {
    sha =
      cleanText(
        fileInfo.sha
      );

    const existingBytes =
      await getExistingGitHubBytes(
        config,
        fileInfo
      );

    if (
      existingBytes &&
      bytesEqual(
        existingBytes,
        newBytes
      )
    ) {
      return {
        ok: true,

        path:
          repoPath,

        action:
          "skipped",

        html_url:
          cleanText(
            fileInfo.html_url
          ),

        commit_sha:
          "",
      };
    }

    action =
      "updated";
  }

  const body = {
    message:
      commitMessage,

    content:
      encodeUtf8Base64(
        content
      ),

    branch:
      config.branch,
  };

  if (sha) {
    body.sha =
      sha;
  }

  const response =
    await fetch(
      contentsUrl(
        config,
        repoPath
      ),
      {
        method:
          "PUT",

        headers: {
          ...githubHeaders(
            config
          ),

          "content-type":
            "application/json",
        },

        body:
          JSON.stringify(
            body
          ),
      }
    );

  if (!response.ok) {
    const detail =
      await response.text();

    throw new Error(
      `GitHub PUT failed for ${repoPath}: ` +
      `${response.status} ${detail.slice(0, 700)}`
    );
  }

  const payload =
    await response.json();

  return {
    ok: true,

    path:
      repoPath,

    action,

    html_url:
      cleanText(
        payload.content
          ?.html_url
      ),

    commit_sha:
      cleanText(
        payload.commit
          ?.sha
      ),
  };
}


function currentPublishDate() {
  const now =
    new Date();

  const year =
    String(
      now.getUTCFullYear()
    );

  const month =
    String(
      now.getUTCMonth() +
      1
    ).padStart(
      2,
      "0"
    );

  const day =
    String(
      now.getUTCDate()
    ).padStart(
      2,
      "0"
    );

  return (
    year +
    month +
    day
  );
}


async function getNextBatchPath(
  config
) {
  const directory =
    "data/publish_batches";

  const url =
    new URL(
      contentsUrl(
        config,
        directory
      )
    );

  url.searchParams.set(
    "ref",
    config.branch
  );

  const response =
    await fetch(
      url,
      {
        headers:
          githubHeaders(
            config
          ),
      }
    );

  let rows = [];

  if (
    response.status !==
    404
  ) {
    if (!response.ok) {
      const detail =
        await response.text();

      throw new Error(
        "Unable to inspect GitHub publish batches: " +
        `${response.status} ${detail.slice(0, 400)}`
      );
    }

    rows =
      await response.json();

    if (
      !Array.isArray(
        rows
      )
    ) {
      rows = [];
    }
  }

  const dateText =
    currentPublishDate();

  const pattern =
    new RegExp(
      `^publish_batch(\\d{2})_${dateText}\\.csv$`,
      "i"
    );

  let maxNumber =
    0;

  for (
    const row
    of rows
  ) {
    const match =
      cleanText(
        row.name
      ).match(
        pattern
      );

    if (!match) {
      continue;
    }

    maxNumber =
      Math.max(
        maxNumber,
        Number(
          match[1]
        )
      );
  }

  const nextNumber =
    maxNumber +
    1;

  return (
    `${directory}/` +
    `publish_batch${String(
      nextNumber
    ).padStart(
      2,
      "0"
    )}_${dateText}.csv`
  );
}


export function getGitHubPublishStatus(
  env
) {
  const config =
    getGitHubConfig(
      env
    );

  return {
    configured:
      Boolean(
        config.token &&
        config.owner &&
        config.repo
      ),

    token_present:
      Boolean(
        config.token
      ),

    owner:
      config.owner,

    repo:
      config.repo,

    branch:
      config.branch,
  };
}


export async function publishWebsiteFilesToGitHub(
  env,
  {
    files,
    commitMessage,
  }
) {
  const config =
    requireGitHubConfig(
      env
    );

  const cleanCommitMessage =
    cleanText(
      commitMessage
    ) ||
    "Update corrosion map website dataset";

  const publishFiles = {
    ...files,
  };

  /*
   * Preserve the legacy dated Site snapshot.
   */
  const sitesCsv =
    publishFiles[
      "data/sites.csv"
    ];

  if (!sitesCsv) {
    throw new Error(
      "Generated website package does not contain data/sites.csv."
    );
  }

  const batchPath =
    await getNextBatchPath(
      config
    );

  publishFiles[
    batchPath
  ] =
    sitesCsv;

  const uploads = [];

  const entries =
    Object.entries(
      publishFiles
    );

  for (
    let index = 0;
    index < entries.length;
    index += 1
  ) {
    const [
      repoPath,
      content,
    ] =
      entries[index];

    try {
      const result =
        await putGitHubFile(
          config,
          {
            repoPath,
            content,
            commitMessage:
              cleanCommitMessage,
          }
        );

      uploads.push(
        result
      );

    } catch (error) {
      return {
        ok: false,

        partial:
          uploads.length > 0,

        changed_count:
          uploads.filter(
            (row) =>
              row.action ===
                "created" ||
              row.action ===
                "updated"
          ).length,

        skipped_count:
          uploads.filter(
            (row) =>
              row.action ===
              "skipped"
          ).length,

        batch_path:
          batchPath,

        uploads,

        failed_path:
          repoPath,

        error:
          error?.message ||
          String(error),
      };
    }
  }

  const changedCount =
    uploads.filter(
      (row) =>
        row.action ===
          "created" ||
        row.action ===
          "updated"
    ).length;

  const skippedCount =
    uploads.filter(
      (row) =>
        row.action ===
        "skipped"
    ).length;

  return {
    ok: true,

    partial:
      false,

    changed_count:
      changedCount,

    skipped_count:
      skippedCount,

    batch_path:
      batchPath,

    uploads,
  };
}